m2cgen/interpreters/f_sharp/interpreter.py

```diff
@@ -0,0 +1,120 @@
+import os
+
+from m2cgen import ast
+from m2cgen.interpreters import mixins, utils
+from m2cgen.interpreters.f_sharp.code_generator import FSharpCodeGenerator
+from m2cgen.interpreters.interpreter import ToCodeInterpreter
+
+
+class FSharpInterpreter(ToCodeInterpreter,
+                        mixins.LinearAlgebraMixin,
+                        mixins.BinExpressionDepthTrackingMixin):
+
+    # Too long lines causes F# compiler to crash with
+    # error FS0193 : internal error :
+    # Specified argument was out of the range of valid values.
+    # (Parameter 'value').
+    # Refer to https://github.com/dotnet/fsharp/issues/3866.
+    bin_depth_threshold = 250
+
+    supported_bin_vector_ops = {
+        ast.BinNumOpType.ADD: "addVectors",
+    }
+
+    supported_bin_vector_num_ops = {
+        ast.BinNumOpType.MUL: "mulVectorNumber",
+    }
+
+    abs_function_name = "abs"
+    exponent_function_name = "exp"
+    logarithm_function_name = "log"
+    log1p_function_name = "log1p"
+    sqrt_function_name = "sqrt"
+    tanh_function_name = "tanh"
+
+    with_log1p_expr = False
+
+    def __init__(self, indent=4, function_name="score", *args, **kwargs):
+        self.indent = indent
+        self.function_name = function_name
+
+        cg = FSharpCodeGenerator(indent=indent)
+        super().__init__(cg, *args, **kwargs)
+
+    def interpret(self, expr):
+        self._cg.reset_state()
+        self._reset_reused_expr_cache()
+
+        args = [(True, self._feature_array_name)]
+        func_name = self.function_name
+
+        with self._cg.function_definition(name=func_name, args=args):
+            last_result = self._do_interpret(expr)
+            self._dump_cache()
+            self._cg.add_code_line(last_result)
+
+        if self.with_linear_algebra:
+            filename = os.path.join(
+                os.path.dirname(__file__), "linear_algebra.fs")
+            self._cg.prepend_code_lines(utils.get_file_content(filename))
+
+        if self.with_log1p_expr:
+            filename = os.path.join(
+                os.path.dirname(__file__), "log1p.fs")
+            self._cg.prepend_code_lines(utils.get_file_content(filename))
+
+        return self._cg.finalize_and_get_generated_code()
+
+    def interpret_if_expr(self, expr, if_code_gen=None, **kwargs):
+        if if_code_gen is None:
```

Comment:
This begins to look like a common pattern among interpreters for functional languages. We should consider generalizing this at some point.